metaflow/plugins/metadata/local.py

```diff
@@ -159,6 +160,40 @@ def _get_object_internal(cls, obj_type, obj_order, sub_type, sub_order, filters=
                 result.append(LocalMetadataProvider._read_json_file(self_file))
         return MetadataProvider._apply_filter(result, filters)
 
+    @classmethod
+    def _perform_operations_internal(cls, operations):
+        to_return = []
+        for op in operations:
+            if op.op_type != 'tags':
+                raise ValueError("Only tag operations are supported")
+            # Get the object we need to update
+            obj_path = op.id.split('/')
+            obj_to_update = cls.get_object(
+                op.object_type, 'self', None, *obj_path)
+            if obj_to_update:
+                # We can perform the operation now
+                if op.operation == 'add':
+                    # We can only add tags that are not already system_tags.
+                    # We also never add duplicate tags to view tags as a set
+                    if op.args['tag'] not in obj_to_update['system_tags'] and \
+                            op.args['tag'] not in obj_to_update['tags']:
+                        obj_to_update['tags'].append(op.args['tag'])
+                elif op.operation == 'remove':
+                    # Remove all instances in case there are duplicates (from
+                    # older MF versions)
+                    obj_to_update['tags'] = [
+                        x for x in obj_to_update['tags'] if x != op.args['tag']]
+                else:
+                    raise TaggingException(msg='Invalid operation %s' % op.operation)
+            else:
+                raise TaggingException(msg='Cannot find object %s' % op.id)
+            # Here, all went well
+            file_path = os.path.join(
+                LocalMetadataProvider._get_metadir(*obj_path), '_self.json')
+            LocalMetadataProvider._dump_json_to_file(file_path, obj_to_update, True)
```

Comment:
the current way of storing tags in a JSON (`_dump_json_to_file`) is not amenable to concurrent operations at all. For instance, if you run two `tag add` operations concurrently, it is possible (likely) that the second tag disappears silently.

Two possible solutions:
1. `flock` the JSON: Flock, read tags, modify tags, write tags, funlock
2. store tags as separate files